### ISSUE TITLE
[v6.0] Accumulate ErgoTree rewriting cost when handling Deserialize operations

### ISSUE THREAD

**Issue by aslesarenko** [ScorexFoundation/sigmastate-interpreter#846]
### Current implementation
In order to execute DeserializeContext operation, the ErgoTree is traversed once. 
During the traverse the DeserializeContext node is replaced by the result of deserialization.
It is done [here](https://github.com/ScorexFoundation/sigmastate-interpreter/blob/f100992f3d2889bdf306b84275b6480e3c819975/sigmastate/src/main/scala/sigmastate/interpreter/Interpreter.scala#L346)

The `deserializeSubstitutionCost` is calculated and accumulated to the new `context1` as `initCost`, however 
the `context1` is erroneously not used in `applyDeserializeContextJITC`.

### What should be done
The `context1` should be used in `applyDeserializeContextJITC` instead of `context`.

**Comment by SethDusek** [comment]
I would like to attempt this.

**Comment by aslesarenko** [comment]
@SethDusek, yes, this change can only be activated (or rather its execution should be conditional on the currently activated protocol version)
Here is an example how such change has been done for v5.0 https://github.com/ScorexFoundation/sigmastate-interpreter/blob/aae4118fed18f6587413d9a6330e449b05d8d5ad/data/shared/src/main/scala/sigma/ast/trees.scala#L38
You can add a new version constant here https://github.com/ScorexFoundation/sigmastate-interpreter/blob/11c98a3382646095ee943b333a0b5bfa65fa291d/core/shared/src/main/scala/sigma/VersionContext.scala#L37-L36

Call it like this
`val ScriptVersionForV6: Byte = 3`

**Comment by aslesarenko** [comment]
Hi, @SethDusek. 
It shouldn't depend on the test cases, but rather on usability and backwards compatibility. 

1) Regardless of the activated version, dApps can create new ergo tree with v0-v2, and with v3 activated, they should continue to do so.
2) Cost estimation is internal mechanism, and we better keep it simple when possible, which means we can do the same calculation of costs for all versions of ErgoTree. 

So, I think we should choose "it apply to any version of ergotree including 0, 1, and 2 provided this is activated on the network".

**Comment by jozanek** [comment]
@kushti possibly fixed by #938

**Comment by Ergologica** [comment]
This appears to be fixed already. On current master `Interpreter.scala:255` reads:

```scala
applyDeserializeContextJITC(if (VersionContext.current.isV6Activated) {
  context1
} else {
  context
}, prop)
```

which is exactly what this issue asks for: `context1`, carrying the `deserializeSubstitutionCost` computed at line 246, is the one passed in, under the v6 activation guard.

Can this be closed, or is something still outstanding? Asking because it is still listed as an open 200 SigUSD bounty in the ErgoDevs registry, so someone may spend time redoing it.

**Comment by jozanek** [comment]
@Ergologica true
